src/libs/compound-v3/service.ts

```diff
@@ -0,0 +1,348 @@
+        }
+      }
+
+      const [baseToken, ...assets] = await this.getTokens([baseTokenAddress, ...assetAddresses]);
+
+      this._marketMap[marketId] = {
+        cometAddress,
+        baseToken,
+        baseTokenPriceFeed,
+        numAssets,
+        utilization: utilization.toString(),
+        assets: assets.map((asset, i) => ({
+          asset,
+          priceFeed: assetPriceFeeds[i],
+          borrowCollateralFactor: borrowCollateralFactors[i],
+          liquidateCollateralFactor: liquidateCollateralFactors[i],
+        })),
+      };
+    }
+
+    return this._marketMap[marketId];
+  }
+
+  async getAprs(marketId: string) {
+    const { cometAddress, utilization } = await this.getMarket(marketId);
+
+    const iface = logics.compoundv3.Comet__factory.createInterface();
+    const calls: common.Multicall2.CallStruct[] = [
+      {
+        target: cometAddress,
+        callData: iface.encodeFunctionData('getSupplyRate', [utilization]),
+      },
+      {
+        target: cometAddress,
+        callData: iface.encodeFunctionData('getBorrowRate', [utilization]),
+      },
+    ];
+    const { returnData } = await this.multicall2.callStatic.aggregate(calls, { blockTag: this.blockTag });
+
+    const [supplyRate] = iface.decodeFunctionResult('getSupplyRate', returnData[0]);
+    const supplyApr = calcApr(supplyRate);
+
+    const [borrowRate] = iface.decodeFunctionResult('getBorrowRate', returnData[1]);
+    const borrowApr = calcApr(borrowRate);
+
+    return { supplyApr, borrowApr };
+  }
+
+  async getPrices(marketId: string) {
+    const { cometAddress, baseTokenPriceFeed, numAssets, assets } = await this.getMarket(marketId);
+    const customBaseTokenPriceFeed = getCustomBaseTokenPriceFeed(this.chainId, marketId);
+
+    const iface = logics.compoundv3.Comet__factory.createInterface();
+    const calls: common.Multicall2.CallStruct[] = [];
+    if (customBaseTokenPriceFeed) {
+      calls.push({
+        target: cometAddress,
+        callData: iface.encodeFunctionData('getPrice', [customBaseTokenPriceFeed]),
+      });
+    }
+    calls.push({
+      target: cometAddress,
+      callData: iface.encodeFunctionData('getPrice', [baseTokenPriceFeed]),
+    });
+    for (const { priceFeed } of assets) {
+      calls.push({
+        target: cometAddress,
+        callData: iface.encodeFunctionData('getPrice', [priceFeed]),
+      });
+    }
+    const { returnData } = await this.multicall2.callStatic.aggregate(calls, { blockTag: this.blockTag });
+
+    let j = 0;
+
+    let customBaseTokenPrice: BigNumber | undefined;
+    if (customBaseTokenPriceFeed) {
+      [customBaseTokenPrice] = iface.decodeFunctionResult('getPrice', returnData[j]);
+      j++;
+    }
+
+    let price: BigNumber;
+    [price] = iface.decodeFunctionResult('getPrice', returnData[j]);
+    if (customBaseTokenPrice) {
+      price = price.mul(customBaseTokenPrice).div(1e8);
```

Comment:
Is it because the price we get here is in `customBaseToken` unit so we need to multiply by `customBaseTokenPrice`(ETH price)?

Reply:
Yes~